lib/ansible/galaxy/collection/__init__.py

```diff
@@ -174,6 +174,7 @@ def verify_local_collection(
         # since we're not downloading this, just seed it with the value from disk
         manifest_hash = get_hash_from_validation_source(MANIFEST_FILENAME)
     else:
+        assert remote_collection is not None  # mypy thing
```

Comment:
why are we adding an assert here? the norm is to avoid `assert` as it can easily be bypassed by python runtime options

Reply:
Otherwise, mypy "thinks" that it may be `None` and issues an error and it doesn't have `is_concrete_artifact` attribute. It's a common idiom/pattern in projects with the type checks enabled.

This is because the incoming function argument `remote_collection` is typed as `Optional[Candidate]` which means "`None` or `Candidate`". Mypy isn't always able to infer the correct type which calls for a need to use `assert`.

I suppose that a [`cast()`](https://mypy.readthedocs.io/en/stable/casts.html) could be used but it'd require an unconditional `from typing import cast` (py3 stdlib + it's currently not used in runtime + all type annotations still rely on the py2-compatible type comments).

P.S. While ansible-core doesn't yet have this sanity check in the CI, there's a PR I've been working on since refactoring of the dep resolver — it's sorta yet ready and was blocked on the `ansible-test` refactoring but at some point, it'll be possible to enable it and it'd better not produce type errors, at least not for these modules that we've covered with the annotations back then.

Reply:
An alternative to the `assert` would be to use `if remote_collection is None` instead of `if verify_local_only`.